File: versions/C#/MicroQiskitC#/MicroQuiskit/SimulatorBase.cs

namespace Qiskit
{
    public class SimulatorBase
    {
        public virtual ComplexNumber[] Simulate(QuantumCircuit circuit)
        {

            double sum = circuit.ProbabilitySum();

            if (sum> MathHelper.Eps)
            {
                if (sum<1- MathHelper.Eps || sum > 1+ MathHelper.Eps)
                {
                    circuit.Normalize(sum);
                }

                ComplexNumber[] amplitudes = new ComplexNumber[circuit.Amplitudes.Length];

                for (int i = 0; i < amplitudes.Length; i++)
                {
                    amplitudes[i] = circuit.Amplitudes[i];
                }
                return amplitudes;
            }
            else
            {
                //Initialize the all 0 vector
                ComplexNumber[] amplitudes = new ComplexNumber[MathHelper.IntegerPower(2, circuit.NumberOfQubits)];
                amplitudes[0].Real = 1;
                return amplitudes;
            }

        }


        public virtual double[] GetProbabilities(QuantumCircuit circuit)
        {
            //Doing nothing just preparing an array
            double[] probabilities = new double[MathHelper.IntegerPower(2, circuit.NumberOfQubits)];
            return probabilities;
        }


        public virtual double[] GetProbabilities(ComplexNumber[] amplitudes)
        {
            //Calculating the probability from the amplitudes
            double[] probabilities = new double[amplitudes.Length];
            for (int i = 0; i < probabilities.Length; i++)
            {
                probabilities[i] = amplitudes[i].Real * amplitudes[i].Real + amplitudes[i].Complex * amplitudes[i].Complex;
            }

            return probabilities;
        }
    }
}

File: versions/Kotlin/src/simulator/SimulatorBase.kt
package simulator

import model.ComplexNumber
import model.QuantumCircuit
import util.MathConstants
import kotlin.math.pow

abstract class SimulatorBase {

    open fun simulate(circuit: QuantumCircuit): List<ComplexNumber> {
        val sum = circuit.probabilitySum()

        return if (sum > MathConstants.Eps) {
            if (sum < 1 - MathConstants.Eps || sum > 1 + MathConstants.Eps)
                circuit.normalize(sum)

            val amplitudes = arrayListOf<ComplexNumber>()
            circuit.amplitudes.forEach { amplitudes.add(it) }

            amplitudes
        } else {
            val amplitudes = arrayListOf<ComplexNumber>()
            for (i in 0 until circuit.amplitudeLength) {
                amplitudes.add(ComplexNumber(0.0, 0.0))
            }
            amplitudes.first().real = 1.0

            amplitudes
        }
    }

    open fun simulateInPlace(circuit: QuantumCircuit): List<ComplexNumber> {
        val localAmplitudes = arrayListOf<ComplexNumber>()
        val sum = circuit.probabilitySum()

        if (sum > MathConstants.Eps) {
            if (sum < 1 - MathConstants.Eps || sum > 1 + MathConstants.Eps) {
                circuit.normalize(sum)
            }

            circuit.amplitudes.forEach { localAmplitudes.add(it) }
        } else {
            circuit.amplitudes.forEach { _ -> localAmplitudes.add(ComplexNumber(0.0, 0.0)) }
            localAmplitudes.first().real = 1.0
        }

        return localAmplitudes
    }

    open fun getProbabilities(circuit: QuantumCircuit): List<Double> {
        val probabilities = arrayListOf<Double>()
        val length = 2.0.pow(circuit.numberOfQubits).toInt()
        for (i in 0..length) probabilities.add(0.0)
        return probabilities
    }

    open fun getProbabilities(amplitudes: List<ComplexNumber>): List<Double> =
        amplitudes.map { it.real.pow(2) + it.complex.pow(2) }

    open fun calculateProbabilities(amplitudes: List<ComplexNumber>, probabilities: List<Double>): List<Double> {
        val localProbabilities = arrayListOf<Double>()
        amplitudes.forEach { _ -> localProbabilities.add(0.0) }

        localProbabilities.mapIndexed { index, _ ->
            amplitudes[index].real.pow(2) + amplitudes[index].complex.pow(2)
        }

        return localProbabilities
    }

}